src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs

```diff
@@ -0,0 +1,278 @@
+        bool isForwardOlder = await IsAncestorCommit(sourceRepo, forwardSha, backwardSha);
+        bool isBackwardOlder = await IsAncestorCommit(sourceRepo, backwardSha, forwardSha);
+
+        // Commits not comparable
+        if (isBackwardOlder == isForwardOlder)
+        {
+            // TODO: Figure out when this can happen and what to do about it
+            throw new Exception($"Failed to determine which commit of {sourceRepo} is older ({lastForwardFlow.VmrSha}, {lastBackflow.VmrSha})");
+        };
+
+        return isBackwardOlder ? lastForwardFlow : lastBackflow;
+    }
+
+    /// <summary>
+    /// Finds the last backflow between a repo and a VMR.
+    /// </summary>
+    private async Task<Backflow?> GetLastBackflow(NativePath repoPath)
+    {
+        // Last backflow SHA comes from Version.Details.xml in the repo
+        var content = await _fileSystem.ReadAllTextAsync(repoPath / VersionFiles.VersionDetailsXml);
+        SourceDependency? source = _versionDetailsParser.ParseVersionDetailsXml(content).Source;
+        if (source is null)
+        {
+            return null;
+        }
+
+        string lastBackflowVmrSha = source.Sha;
+        string lastBackflowRepoSha = await BlameLineAsync(
+            repoPath / VersionFiles.VersionDetailsXml,
+            line => line.Contains(VersionDetailsParser.SourceElementName) && line.Contains(lastBackflowVmrSha));
+
+        return new Backflow(lastBackflowVmrSha, lastBackflowRepoSha);
+    }
+
+    /// <summary>
+    /// Finds the last forward flow between a repo and a VMR.
+    /// </summary>
+    private async Task<ForwardFlow> GetLastForwardFlow(string mappingName)
+    {
+        IVersionedSourceComponent repoInVmr = _sourceManifest.Repositories.FirstOrDefault(r => r.Path == mappingName)
+            ?? throw new ArgumentException($"No repository mapping named {mappingName} found");
+
+        // Last forward flow SHAs come from source-manifest.json in the VMR
+        string lastForwardRepoSha = repoInVmr.CommitSha;
+        string lastForwardVmrSha = await BlameLineAsync(
+            _vmrInfo.SourceManifestPath,
+            line => line.Contains(lastForwardRepoSha));
+
+        return new ForwardFlow(lastForwardVmrSha, lastForwardRepoSha);
+    }
+
+    /// <summary>
+    /// Compares 2 git commits and returns true if the first one is an ancestor of the second one.
+    /// </summary>
+    private async Task<bool> IsAncestorCommit(NativePath repoPath, string parent, string ancestor)
+    {
+        var result = await _processManager.ExecuteGit(repoPath, ["merge-base", "--is-ancestor", parent, ancestor]);
+
+        if (!string.IsNullOrEmpty(result.StandardError))
+        {
+            result.ThrowIfFailed($"Failed to determine which commit of {repoPath} is older ({parent}, {ancestor})");
+        }
+
+        return result.ExitCode == 0;
+    }
+
+    protected abstract record Codeflow(string SourceSha, string TargetSha)
```

Comment:
> What I don't understand though is how the Codeflow properties RepoSha and VmrSha are mapped to the primary parameters of SourceSha and TargetSha

The child classes implement these abstract properties by defining those via the primary constructor (which makes the parameters public properties).

> I understand it was intentional, but it would make sense to me to define ForwardFlow as ForwardFlow(string RepoSha, string VmrSha) so that the reading direction follows the flow.

A good idea, I will change this so that the order of the parameters represents the direction of the flow.

> These three records are very confusing, especially the naming of the arguments and properties.

This actually helps a lot in the usage where for creating these, it's just simpler to match the parameters by name (`vmrSha`, `repoSha`) and not care about the order of things since we already know which flow we're creating.
But later, in common parts for back and forward flow, it's more readable to use `Source` and `Target` as it simplifies the reading. So that we don't need to do `isBackflow ? flow.RepoSha : flow.VmrSha` all the time.

Reply:
> But later, in common parts for back and forward flow, it's more readable to use Source and Target as it simplifies the reading. So that we don't need to do isBackflow ? flow.RepoSha : flow.VmrSha all the time.

Agree on this.